tests/ping_pong.cpp

```diff
@@ -0,0 +1,72 @@
+#include <fmt/color.h>
+#include <fmt/core.h>
+
+#include <liburing/io_service.hpp>
+#include <string_view>
+auto ping(uio::io_service& service, int read_fd, int write_fd) -> uio::task<> {
+    std::string_view ping = "ping!";
+    std::array<char, 64> buffer;
+    for (int i = 0; i < 20; i++) {
+        int count =
+            co_await service.read(read_fd, buffer.data(), buffer.size(), 0);
+        if (count < 0)
+            uio::panic("Bad read", errno);
+
+        auto recieved = std::string_view(buffer.data(), count);
+
+        fmt::print(
+            fg(fmt::color::cyan) | fmt::emphasis::bold,
+            "ping: Recieved {}\n",
+            recieved);
+        if (recieved != "pong!")
+            uio::panic("Unexpected message", 0);
+
+        co_await service.write(write_fd, ping.data(), ping.size(), 0);
+        co_await service.fsync(write_fd, 0);
+    }
+    close(read_fd);
+    close(write_fd);
+}
+auto pong(uio::io_service& service, int read_fd, int write_fd) -> uio::task<> {
+    std::string_view ping = "pong!";
+    std::array<char, 64> buffer;
+    for (int i = 0; i < 20; i++) {
+        co_await service.write(write_fd, ping.data(), ping.size(), 0);
+        co_await service.fsync(write_fd, 0);
+
+        int count =
+            co_await service.read(read_fd, buffer.data(), buffer.size(), 0);
+        if (count < 0)
+            uio::panic("Bad read", errno);
```

Comment:
io_uring IO functions don't set the global errno, they return -errno on error. See panic_on_error